tsrc/git.py

```diff
@@ -113,13 +113,71 @@ def reset(repo, ref):
     run_git(repo, "reset", "--hard", ref)
 
 
-def get_status(working_path):
+def get_complete_status(working_path):
     _, out = run_git(working_path, "status", "--porcelain", raises=False)
+
+    untracked = 0
+    staged = 0
+    not_staged = 0
+    added = 0
+
     for line in out.splitlines():
         if line.startswith("??"):
-            return "untracked files"
+            untracked += 1
         if line.startswith(" M"):
-            return "modified files"
+            staged += 1
+        if line.startswith(" .M"):
+            not_staged += 1
         if line.startswith("A "):
-            return "non-committed files"
-    return "clean"
+            added += 1
+
+    return untracked, staged, not_staged, added
+
+
+def commit_string(symbol, number):
+    commit = '%s%s commit' % (symbol, number)
+    if number > 1:
+        commit += 's'
+    return commit
+
+
+def git_remote_status(working_path):
+    _, ahead_rev = run_git(working_path, "rev-list", "@{upstream}..HEAD", raises=False)
+    ahead = len(ahead_rev.splitlines())
+
+    _, behind_rev = run_git(working_path, "rev-list", "HEAD..@{upstream}", raises=False)
+    behind = len(behind_rev.splitlines())
+
+    ahead_symbol = "↑"
+    behind_symbol = "↓"
+
+    status = ''
+
+    if ahead > 0:
+        if behind == 0:
+            status = commit_string(ahead_symbol, ahead)
+        else:  # i.e. if behind > 0
+            status = commit_string(ahead_symbol, ahead)
+            status += ', '
+            status += commit_string(behind_symbol, behind)
+    else:  # if ahead == 0
+        if behind > 0:
+            status = commit_string(behind_symbol, behind)
+
+    return status
+
+
+def get_status(working_path):
```

Comment:
I think we can find better architecture there.

Maybe instead of having a tuples as the return value of the various get_() function we can move the `Status` class inside`tsrc.git`, and then move the logic of transforming the status to a string to `tsrc.cli.status`:

Something like this:

```python
# in tsrc.git.py

class GitStatus:
    def __init__(self, working_path):
        self.working_path = working_path
        self.untracked = 0
        self.staged = 0
        self.ahead = 0
        self.behind = 0
        self.dirty = False
        self.branch = None

    def update(self):
        self.update_branch()
        self.update_remote_status()
        self.update_worktree_status()

    def update_branch(self):
        # No-op if get_current_branch throws
        ...

    def update_remote_status(self):       
        ...
    def update_worktree_status(self):
         ....


def get_status(working_path):
    status = GitStatus(working_path)
    status.update()


# in tsrc.cli.status

def describe(git_status):
    # Return a list of tokens suitable for ui.info()
    ...
    res = []
    res += describe_branch(git_status)    # add (ui.green, branch_name) or (ui.red, not on any branch)
    res += describe_position(git_status)  # add (up, commits)
    res += describe_dirty(git_status)     # add (ui.red, "dirty")
    return res


# collect_statuses is gone

def display_statuses(workspace):
    for repo in workspace.repos:
        status = tsrc.git.get_status(full_path)
        ui.info(*describe(status))
```

This new architecture has the following benefits:

* The 'errors' in `tsrc status` are gone. The fact that the repo is not on any branch should not prevent us
  from displaying whether the repository is clean or not

* The `Status` class in `tsrc.cli.status` is gone. It was mixing two difference pieces of data: the state of the repository, and its path in the workspace (the `src` attribute).

* We can extend the `GitStatus` class with new information if we need it elsewhere, but we are free to change the logic on how the status gets displayed independently.

Does this makes sense to you ?

Now, I realize this is quite a big refactoring, so if you need help, don't hesitate to ask